### ISSUE TITLE
Associations break life cycle method semantics

### ISSUE THREAD

**Issue by oneHappyCanuck** [balderdashy/sails#2735]
I'm noticing that models with associations trigger update life cycle methods in the context of a create. For example if I have Model A with one to many associations to model B and I do A.create (with associations populated) I see the following life cycle method invocations:

A.beforeCreate
A.beforeUpdate
A.afterUpdate
B.beforeCreate
B.afterCreate
A.afterCreate

I understand the rationale behind this, given the need to modify the model attributes to implement associations, but invoking the update life cycle methods in the process of wiring up associations changes the semantics of the life cycle methods substantially.

For example I have distinct validation requirements for create and update scenarios, but when using associations I lose the ability to differentiate between create and update scenarios given that before/after Update is called in the context of a create *and* update.

I can work around this by moving responsibilities around - for example by invoking model validation in the context of a controller or service rather than using the life cycle methods and by issuing events at the controller or service rather than at the model - but it would be *so* much cleaner if the semantics of the life cycle method were preserved. 

Wondering if I'm missing something?

**Comment by tjwebb** [comment]
> A.beforeCreate
A.beforeUpdate
A.afterUpdate
B.beforeCreate
B.afterCreate
A.afterCreate

To me, this looks appropriate and correct. Can you provide some more detail on your specific use case?

**Comment by oneHappyCanuck** [comment]
Sure, 
We have model validation that requires async lookups, which falls outside what can be achieved using Sails declarative validation, so we implement this validation via Model.beforeCreate() and Model.beforeUpdate(). We have separate validation requirements for create and update scenarios so the implementation is distinct.  
When using associations Model.create() causes calls to Model.beforeCreate() *and* Model.beforeUpdate() with the values provided to Model.create(), and our beforeUpdate() validation fails because some of the values provided are inappropriate for an update and we cannot differentiate between a beforeUpdate() call in the context of Model.update() and a beforeUpdate() call invoked in the context of Model.create() with associations.  
From my perspective Sails does a great job of defining a small number of highly cohesive and loosely coupled abstractions, and having Model.beforeUpdate() called in the context of a create operation in this fashion blurs the semantics of your api's. IMO the best solution would be to simply suppress Model.before/afterUpdate calls when they are triggered by the implementation details of setting up associations in the context of Model.create().
My .02..  
Great framework. 
Cheers, 



  
Date: Sun, 22 Mar 2015 19:40:28 -0700

**Comment by sailsbot** [comment]
Thanks for posting, @oneHappyCanuck.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
